src/Effect/Ref.purs

```diff
@@ -2,6 +2,38 @@
 -- |
 -- | _Note_: `Control.Monad.ST` provides a _safe_ alternative to `Ref` when
 -- | mutation is restricted to a local scope.
+-- |
+-- | You'll notice that all of the functions that operate on a `Ref`
+-- | (e.g. `new`, `modify`, `modify_`, `read`) return an `Effect (Ref s)`
+-- | and not a `Ref s`. Here's what would happe if the `Ref s` was not
+-- | wrapped in an `Effect`:
+-- | 1. In some situations, one could violate referential transparency.
+-- | 2. One can violate the type system by giving a `Ref` a polymorphic type
+-- |    and then specializing it after the fact.
+-- |
+-- | In regards to the first, consider the below example:
```

Comment:
For the topic of why using Refs is effectful in general, I think it will probably be more persuasive to use a simpler example where a Ref is created, read from, written to, and then read from again: that's likely to be easier to understand than this example, which is a bit more subtle.